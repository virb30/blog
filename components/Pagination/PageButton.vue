<template>
    <button type="button" v-bind="$attrs" :disabled="disabled || active"
        class="h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600"
        :class="buttonClasses" @click="$emit('click', $event)">
        <slot></slot>
    </button>
</template>

<script lang="ts" setup>
import { computed } from 'vue';

defineEmits(['click'])

const props = defineProps({
    active: {
        type: Boolean,
        default: false,
    },
    disabled: {
        type: Boolean,
        default: false,
    },
})

const buttonClasses = computed(() => (
    {
        'opacity-40 cursor-not-allowed ': props.disabled,
        'bg-indigo-500 dark:bg-indigo-500 cursor-default text-gray-50': props.active,
        'hover:opacity-80 cursor-pointer': !props.disabled && !props.active,
    }
));
</script>
