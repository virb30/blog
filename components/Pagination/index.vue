<template>
	<div class="flex flex-col items-center my-12">
		<div class="flex">
			<page-button
				type="button"
				:disabled="!pagination.prev"
				@click="prevPage"
			>
				<FontAwesomeIcon :icon="['fas', 'chevron-left']" />
			</page-button>
			<page-button
				v-for="page in pagination.pages"
				:key="page"
				:active="page === pagination.page"
				@click="toPage(page)"
			>
				{{ page }}
			</page-button>
			<page-button
				:disabled="!pagination.next"
				@click="nextPage"
			>
				<FontAwesomeIcon :icon="['fas', 'chevron-right']" />
			</page-button>
		</div>
	</div>
</template>

<script lang="ts">
import Vue from 'vue';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { mapActions, mapState } from 'vuex';
import { Pagination } from '@tryghost/content-api';
import PageButton from './PageButton.vue';

library.add(faChevronLeft, faChevronRight);

interface Methods {
	prevPage(): void
	toPage(page: number): void
	nextPage(): void
	fetchPosts(page: number): void
}

interface Computed {
	pagination: Pagination,
}

export default Vue.extend<unknown, Methods, Computed, unknown>({
	components: { FontAwesomeIcon, PageButton },
	computed: {
		...mapState('Blog', ['pagination']),
	},
	methods: {
		...mapActions('Blog', ['fetchPosts']),
		nextPage () {
			if (this.pagination.next) {
				this.toPage(this.pagination.next);
			}
		},
		prevPage () {
			if (this.pagination.prev) {
				this.toPage(this.pagination.prev);
			}
		},
		toPage (page: number) {
			this.fetchPosts(page);
		},
	},
});
</script>
