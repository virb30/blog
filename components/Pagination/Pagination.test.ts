import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import Pagination from './index.vue';
import PageButton from './PageButton.vue';

describe('Pagination Component', () => {
    describe('render', () => {
        it('should render childs', () => {
            const wrapper = mount(Pagination, {
                global: {
                    stubs: {
                        'font-awesome-icon': true
                    }
                },
                props: {
                    pagination: {
                        page: 1,
                        pages: 2,
                        next: 2,
                        prev: null,
                    },
                },
            });

            const buttons = wrapper.findAll('button');

            expect(wrapper.html()).toMatchSnapshot();
            expect(buttons.length).toBe(4);
        });
    });

    describe('navigation', () => {
        it('should navigate to next page', async () => {
            const wrapper = mount(Pagination, {
                global: {
                    stubs: {
                        'font-awesome-icon': true
                    }
                },
                props: {
                    pagination: {
                        page: 1,
                        pages: 2,
                        next: 2,
                        prev: null,
                    },
                },
            });

            const pageButtons = wrapper.findAllComponents(PageButton);

            await pageButtons?.at(3)?.trigger('click')

            await wrapper.vm.$nextTick()

            expect(wrapper.emitted()['change-page'][0]).toEqual([2])
        });

        it('should not navigate', async () => {
            const wrapper = mount(Pagination, {
                global: {
                    stubs: {
                        'font-awesome-icon': true
                    }
                },
                props: {
                    pagination: {
                        page: 1,
                        pages: 1,
                        next: null,
                        prev: null,
                    },
                },
            });

            const pageButtons = wrapper.findAllComponents(PageButton);

            await pageButtons?.at(0)?.trigger('click');
            await wrapper.vm.$nextTick()
            expect(wrapper.emitted()['change-page']).toBeUndefined();

            await pageButtons?.at(2)?.trigger('click');
            await wrapper.vm.$nextTick()
            expect(wrapper.emitted()['change-page']).toBeUndefined();
        });

        it('should navigate to prev page', async () => {
            const wrapper = mount(Pagination, {
                global: {
                    stubs: {
                        'font-awesome-icon': true
                    }
                },
                props: {
                    pagination: {
                        page: 2,
                        pages: 2,
                        next: null,
                        prev: 1,
                    },
                },
            });

            const pageButtons = wrapper.findAllComponents(PageButton);

            await pageButtons?.at(0)?.trigger('click');
            await wrapper.vm.$nextTick()
            expect(wrapper.emitted()['change-page'][0]).toEqual([1]);
        });

        it('should navigate to arbitrary page', async () => {
            const wrapper = mount(Pagination, {
                global: {
                    stubs: {
                        'font-awesome-icon': true
                    }
                },
                props: {
                    pagination: {
                        page: 1,
                        pages: 3,
                        next: 2,
                        prev: null,
                    },
                },
            });

            const toPage = vi.spyOn(wrapper.vm, 'toPage');

            const pageButtons = wrapper.findAllComponents(PageButton);

            await pageButtons?.at(3)?.trigger('click');
            await wrapper.vm.$nextTick()
            expect(wrapper.emitted()['change-page'][0]).toEqual([3]);
        });
    });
});
