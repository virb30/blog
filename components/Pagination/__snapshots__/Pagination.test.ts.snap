// Vitest Snapshot v1

exports[`Pagination Component > render > should render childs 1`] = `
"<div class=\\"flex flex-col items-center my-12\\">
  <div class=\\"flex\\"><button type=\\"button\\" disabled=\\"\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 opacity-40 cursor-not-allowed\\">
      <font-awesome-icon-stub icon=\\"fas,chevron-left\\"></font-awesome-icon-stub>
    </button><button type=\\"button\\" disabled=\\"\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 bg-indigo-500 dark:bg-indigo-500 cursor-default text-gray-50\\">1</button><button type=\\"button\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 hover:opacity-80 cursor-pointer\\">2</button><button type=\\"button\\" class=\\"h-8 w-8 flex justify-center items-center border border-gray-400 bg-gray-300 dark:bg-gray-600 hover:opacity-80 cursor-pointer\\">
      <font-awesome-icon-stub icon=\\"fas,chevron-right\\"></font-awesome-icon-stub>
    </button></div>
</div>"
`;
